File: nc/practice/recursion/tower_of_hanoi.c
/** Name -> Abhay Shanker Pathak
Date -> 2020-05-05
Description -> recursion problem, tower of hanoi
Email -> abhaysp9955@gmail.com
Github -> https://github.com/CoolAbhayS
Last Edit -> 2020-05-05 */


#include<stdio.h>

int TOH(int ,char ,char ,char);

static int count = 0;

// main function
// <<<
int main(int argc, char* argv[]) {
	int disk_count, cycle;
	char first='a', second='b', third='c';
	printf("Enter the number of disks: ");
	scanf("%d", &disk_count);
	cycle = TOH(disk_count, first, second, third);
	printf("Total number of cycles: %d\n", cycle);
	return 0;
}
// >>>

// function for tower of hanoi
// <<<
int TOH(int num, char a, char b, char c) {
	if (num > 0) {
		TOH(num - 1, a, c, b);
		printf("Moving last disk from %c to %c\n", a, c);
		count += 1;
		TOH(num - 1, b, a, c);
	}
	return count;
}
// >>>

File: npython/practice/recursion/tower_of_hanoi.py
#!/usr/bin/env python

"""
Author: Abhay Shanker Pathak
Date: 2020-05-05
Description: recursive function, tower of hanoi
Email: abhaysp9955@gmail.com
Last Edit: 2020-05-05
"""

PROMPT = '> '
cycle = 0


# function for tower of hanoi
# <<<
def toh(count, first, second, third):
    '''function to get tower of hanoi count moves'''
    global cycle
    if count > 0:
        toh(count - 1, first, third, second)
        print('Moving last disk from ', first, ' to ', third)
        cycle += 1
        toh(count - 1, second, first, third)
# >>>


# main function
# <<<
def main():
    '''Entering in main function'''
    print('Enter the disk count: ')
    disk_count = int(input(PROMPT))
    toh(disk_count, 1, 2, 3)
    print('Total cycle is: ', cycle)
# >>>


if __name__ == "__main__":
    main()
